### PR TITLE
add permissions

### LINKED ISSUE #17: permissions
idea: add them to viewinfo and use that as source

### PR DESCRIPTION
**Pull request by RaphaelKimmig**
fixes #17

### CODE REVIEW

File: src/beam/components.py [hunk 1]
```diff
@@ -39,6 +43,17 @@ def get_arguments(cls) -> Set[str]:
...
+    def has_perm(self, user, obj=None):
+        if self.permission is None:
+            return True
+        if not user:
+            return False
+        if callable(self.permission):
+            return self.permission(user, obj=obj)
+        # FIXME support for object permissions
+        # the ModelBackend returns False as soon as we supply an obj which gives us a bad default :-(
+        return user.has_perm(self.permission)
```
Comment (RaphaelKimmig) [inline src/beam/components.py:55]: Gedanken dazu @HannesWaechter ?

Comment (RaphaelKimmig) [reply src/beam/components.py:55]: implementation in ModelAdmin:
```
    def has_change_permission(self, request, obj=None):
        """
        Return True if the given request has permission to change the given
        Django model instance, the default implementation doesn't examine the
        `obj` parameter.

        Can be overridden by the user in subclasses. In such case it should
        return True if the given request has permission to change the `obj`
        model instance. If `obj` is None, this should return True if the given
        request has permission to change *any* object of the given type.
        """
        opts = self.opts
        codename = get_permission_codename('change', opts)
        return request.user.has_perm("%s.%s" % (opts.app_label, codename))
```

File: tests/test_contrib_reversion.py [hunk 2]
```diff
@@ -156,13 +156,12 @@ def test_revision_is_visible_in_list():
```
Comment (HannesWaechter) [inline tests/test_contrib_reversion.py:156]: test_revert_to_previous_version doesn't need perms?
besides that maybe some user_without_perms-tests would be nice

File: tests/test_views.py [hunk 3]
```diff
@@ -143,7 +176,8 @@ def test_create_with_inlines(client):
```
Comment (HannesWaechter) [inline tests/test_views.py:176]: Maybe we want some user_without_permission-tests here too?

File: src/beam/views.py [hunk 4]
```diff
@@ -50,6 +50,15 @@ def fields(self):
...
 
+    def dispatch(self, request, *args, **kwargs):
+        try:
+            obj = self.get_object()
+        except AttributeError:
+            obj = None
+        if not self.component.has_perm(request.user, obj):
+            raise PermissionDenied("You shall not pass")
+        return super().dispatch(request, *args, **kwargs)
+
```
[the remark this answers is not in the document]
Comment (RaphaelKimmig) [reply src/beam/views.py:61]: yep, good point!

File: src/beam/components.py [hunk 5]
```diff
@@ -39,6 +43,17 @@ def get_arguments(cls) -> Set[str]:
...
 
+    def has_perm(self, user, obj=None):
+        if self.permission is None:
+            return True
+        if not user:
+            return False
+        if callable(self.permission):
+            return self.permission(user, obj=obj)
+        # FIXME support for object permissions
+        # the ModelBackend returns False as soon as we supply an obj which gives us a bad default :-(
```
[the remark this answers is not in the document]
Comment (RaphaelKimmig) [reply src/beam/components.py:54]: It does and will always return False if it is passed

File: src/beam/components.py [hunk 6]
```diff
@@ -64,9 +79,10 @@ def __init__(
         model=None,
         inline_classes=None,
         link_layout=None,
+        name=None,
+        url_name=None,
```
[the remark this answers is not in the document]
Comment (RaphaelKimmig) [reply src/beam/components.py:83]: That happens in the base class, we only provide defaults here.

File: src/beam/components.py [hunk 7]
```diff
@@ -21,10 +21,14 @@ def __init__(
 
         self.name = name
         self.verbose_name = verbose_name or self.name
-        self.permission = permission  # FIXME string / list / callable
         self.url_name = url_name
         self.url_kwargs = url_kwargs
 
+        if isinstance(permission, str):
+            self.permission = permission.format(component=self)
```
[the remark this answers is not in the document]
Comment (RaphaelKimmig) [reply src/beam/components.py:28]: That would mean every that every component has to have a model

File: tests/test_contrib_reversion.py (hunk 2 above)
Comment (RaphaelKimmig) [reply tests/test_contrib_reversion.py:156]: Good catch; I'll look into that. Also I'll add no-permission tests for all views

File: src/beam/components.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (HannesWaechter) [reply src/beam/components.py:28]: maybe we look if there is a model?

**Comment by HannesWaechter** [comment]
> One last thing: what about permissions for inlines?

The ModelAdmin checks add, change and delete permissions for inlines:
```
    def get_inline_instances(self, request, obj=None):
        inline_instances = []
        for inline_class in self.inlines:
            inline = inline_class(self.model, self.admin_site)
            if request:
                if not (inline.has_add_permission(request) or
                        inline.has_change_permission(request, obj) or
                        inline.has_delete_permission(request, obj)):
                    continue
                if not inline.has_add_permission(request):
                    inline.max_num = 0
            inline_instances.append(inline)

        return inline_instances
```

**Comment by RaphaelKimmig** [comment]
> One last thing: what about permissions for inlines?

Let's do this in a separate mr

**Pull request merged**